technology-considerations/graphql-prototypes.md

```diff
@@ -0,0 +1,178 @@
+
+We investigated two approaches: AWS AppSync and the Scala Sangria library. We've
+decided on continuing with the Sangria approach because the development and
+deployment environments are much easier to set up and maintain.
+
+[api-options]: https://github.com/nationalarchives/tdr-dev-documentation/pull/4
+
+## Context
+
+TDR needs to store information on the state of consignments of files as they are
+transferred to The National Archives. For example, it will store the original
+file paths of files, along with their checksums and the result of virus scans.
+The files themselves will be stored in [AWS S3][s3], and we've decided to [store
+the rest of the data in a relational DB][db-choice].
+
+Several systems need to be able to read and update the data:
+
+- The TDR frontend (both in the browser and in server-side code)
+- Background tasks that analyse uploaded files
+- The export process which will save the file metadata
+
+The API layer will provide a consistent interface for these clients to access
+the data.
+
+Whatever approach we take for the API, we would prefer to write any business
+logic in Scala. The team developing TDR uses Scala and Java in our other
+server-side applications, and we've chosen to keep using Scala for consistency.
+
+[s3]: https://aws.amazon.com/s3/
+[db-choice]: https://github.com/nationalarchives/tdr-dev-documentation/pull/7
+
+## AppSync approach
+
+[AWS AppSync][appsync] is a platform specifically for building GraphQL APIs, and
+it can fetch data from multiple data sources including AWS RDS databases.
+
+In AppSync, [GraphQL resolvers][resolvers] can be implemented in one of two
+ways. Simple resolvers, which are just a wrapper for a single database table,
+can use [Apache Velocity templates][velocity] to map between the GraphQL request
+and the database. More complex resolvers, which implement business or
+authorization logic, can be written as [pipeline resolvers][pipeline-resolvers],
+which call arbitrary AWS Lambda functions.
+
+We investigated this option because it sounded like it would save us setup and
+deployment time, because request handling and database connections are built
+into AppSync. In practice, though, we ran into a lot of problems.
+
+[appsync]: https://aws.amazon.com/appsync/
+[resolvers]: https://graphql.org/learn/execution/
+[velocity]: https://docs.aws.amazon.com/appsync/latest/devguide/resolver-mapping-template-reference-programming-guide.html
+[pipeline-resolvers]: https://docs.aws.amazon.com/appsync/latest/devguide/pipeline-resolvers.html
+
+### Local development
+
+It was fairly straightforward to set up an [AppSync
+prototype][tdr-appsync-prototype] based on the [Serverless
+Framework][serverless]. This connected to DynamoDB and just used Velocity to map
+the data.
+
+As mentioned above, we wanted to check that we could write business logic in
+Scala. [Serverless supports Lambdas written in Java][serverless-java] (which
```

Comment:
IntelliJ https://plugins.jetbrains.com/plugin/11349-aws-toolkit can be used. https://docs.aws.amazon.com/toolkit-for-jetbrains/latest/userguide/key-tasks.html#key-tasks-install 
Requires 2018.3 or later so need licence extension

Reply:
Do you know if it supports AppSync? I can't see any mention of it.

Reply:
Don't know if it supports appsync. Just  AWS toolkit could be used for our Lambda development